### DISCUSSION TITLE
Updating 'Row' with model

### DISCUSSION THREAD

**Discussion by danbolinson-bt** [anvilistas/tabulator#70] [Q&A]
I am using a model and that model is updateable by 'formatters' in the table. Other properties depend on attributes that can be changed.

I am trying to have Tabulator update the Row after an update (basically to ```refresh_data_bindings``` on the tabulator row object).

I can't seem to find a way to 'force' this to occur in Tabulator. I have tried:

```
        cell.get_row().update(cell.get_row().get_model())
        cell.get_table().redraw()
        uid = cell.get_row().get_model().uid
        # for some reason cell.get_row().get_index() returns None
        cell.get_table().updateRow(uid, cell.get_row().get_model())
```

None of these cause an update to occur. Clone link:

https://anvil.works/new-build/apps/VA3VTMAVOVTEGOI3/code/forms/Form1#design;code:45:67

I'm considering a messenger to pass 'row updated' messages that other components can listen to, but that feels hacky. Any ideas? Thanks!

**Comment by s-cork** [comment]
The link isn't a clone link - can you send it again?

**Reply by danbolinson-bt** [reply]
Ooops, sorry:
https://anvil.works/build#clone:VA3VTMAVOVTEGOI3=W2EKAGTD23TC2J3SRQBMEJUG

**Reply by s-cork** [reply]
Thanks. Will look at this next week. Maybe needs some work on the internals for updating models.

**Comment by s-cork** [comment]
So the issue here is that the `get_row().update()` function from tabulator takes a dictionary of key value pairs to update its version of the data it holds.
The data it holds will be a javascript object and is a copy of the data from the model object.

Sending the model object isn't what the update function expects and so we get the same problem as before,
the python object is being wrapped in JS world and can't be handled by tabulator to perform the necessary updates.

We have 2 things that we're trying to keep in sync.
- A model object
- row data
When something is changed you'll need to update both.
If you use Tabulator's native editors then the row data will get updated and you'd need to update the model.

One way you might do this more simply is to have an `as_dict` method.
Then you can do
```python
row = cell.get_row()
model = row.get_model()
model.foo = "bar"
row.update(model.as_dict())
```

I'm sure it would be possible to support an API for making this easier, but it's also probably something I don't have time to implement any time soon.

**Reply by lewysigns** [reply]
Thank you ! This is exactly what I needed! Also classes have a built in __dict__ dunder you could you (Tested and verified it works).

**Comment by lewysigns** [comment]
Along the same vein, how would you go about adding a row to your table using models?

**Reply by s-cork** [reply]
That'll be part of the same problem. I'll see if I have a moment this week to add some methods just for models.

**Reply by s-cork** [reply]
@lewysigns - that should just work on the development branch - I'll push a new version

i.e. `self.tabulator.add_data(new_model_instance)`

**Reply by lewysigns** [reply]
Thank you for making time Stu. 

I'll look at the development branch and see how I can contribute next time! ( I was trying to understand )

**Reply by s-cork** [reply]
This should now work on the Production branch too
or `v2.1.6+` if you prefer to tag to a specific version number

**Reply by lewysigns** [reply]
you're a saint.
